src/MgvOfferTaking.sol

```diff
@@ -306,146 +306,97 @@ abstract contract MgvOfferTaking is MgvHasOffers {
-          if (mgvData != "mgv/notExecuted") {
-            // updates best&tick
-            sor.local = dislodgeOffer(pair, sor.offer, sor.local, true);
-          }
+        /* In the market order, we were able to avoid stitching back offers after every `execute` since we knew a continuous segment starting at best would be consumed. Here, we cannot do this optimisation since offers in the `targets` array may be anywhere in the book. So we stitch together offers immediately after each `execute`. */
+        sor.local = dislodgeOffer(pair, sor.offer, sor.local, true);
 
-          /* <a id="internalSnipes/liftReentrancy"></a> Now that the current snipe is over, we can lift the lock on the book. In the same operation we
+        /* <a id="internalSnipes/liftReentrancy"></a> Now that the current snipe is over, we can lift the lock on the book. In the same operation we
         * lift the reentrancy lock, and
         * update the storage
 
         so we are free from out of order storage writes.
         */
-          sor.local = sor.local.lock(false);
-          pair.local = sor.local;
-
-          /* `payTakerMinusFees` keeps the fee in Mangrove, proportional to the amount purchased, and gives the rest to the taker */
-          payTakerMinusFees(mor, sor);
+        sor.local = sor.local.lock(false);
+        pair.local = sor.local;
 
-          /* In an inverted Mangrove, amounts have been lent by each offer's maker to the taker. We now call the taker. This is a noop in a normal Mangrove. */
-          executeEnd(mor, sor);
+        /* No fees are paid since offer execution failed. */
 
-          /* After an offer execution, we may run callbacks and increase the total penalty. As that part is common to market orders and snipes, it lives in its own `postExecute` function. */
-          if (mgvData != "mgv/notExecuted") {
-            postExecute(mor, sor, gasused, makerData, mgvData);
-          }
+        /* In an inverted Mangrove, amounts have been lent by each offer's maker to the taker. We now call the taker. This is a noop in a normal Mangrove. */
+        // FIXME: Is this needed when offer execution failed? I don't think so, but should be covered by tests
+        // executeEnd(mor, sor);
 
-          snipesGot += mor.totalGot;
-          snipesGave += mor.totalGave;
-        }
+        /* After an offer execution, we may run callbacks and increase the total penalty. As that part is common to market orders and snipes, it lives in its own `postExecute` function. */
+        postExecute(mor, sor, gasused, makerData, mgvData);
       }
+
+      bounty = mor.totalPenalty;
+
+      /* Over the course of the snipes order, a penalty reserved for `msg.sender` has accumulated in `mor.totalPenalty`. No actual transfers have occured yet -- all the ethers given by the makers as provision are owned by Mangrove. `sendPenalty` finally gives the accumulated penalty to `msg.sender`. */
+      sendPenalty(bounty);
```

Comment:
Previously we could clean multiple offers and then in the end do the transfer (at cast 21000 gas). We now pay that amount of gas for each transfer, thus upping the needed penalty to cover the cost.

Should gasbase compensate a normal market order taker, or a cleaner? If it is to incentivize cleaning, then that is the important one.

Reply:
I agree -- gasbase should compensate max(market order overhead, cleaning overhead). 

I didn't get the 21k gas reference?

Reply:
Good point!

I think this is a strong reason for moving the multi-clean functionality from `MgvCleaner` to the core. Should be pretty quick to do.

Unless @adhusson objects, I'll go ahead and make the change.

Reply:
Done in [e4cfb66](https://github.com/mangrovedao/mangrove-core/pull/437/commits/e4cfb668bd6ebfb2547952379be91c8c510a1482)